### PR TITLE
[WIP] Provide necessary APIs needed to run nginx.wasm

### PR DESCRIPTION
**Pull request by appcypher**
**What is done:**
* Established base work for emscripten runtime elements
* Implemented some required APIs

**What is in progress:**
* Completing apis and runtime elements necessary for running `nginx.wasm` successfully
* Writing tests for implemented APIs
* Documenting work on APIs

### CODE REVIEW

File: src/webassembly/instance.rs [hunk 1]
```diff
@@ -102,7 +101,6 @@ pub struct Instance {
 /// Contains pointers to data (heaps, globals, tables) needed
 /// by Cranelift.
 /// NOTE: Rearranging the fields will break the memory arrangement model
-#[repr(C)]
```
Comment (syrusakbary) [inline src/webassembly/instance.rs]: Good catch. Not sure why we had it twice.

File: src/webassembly/memory.rs [hunk 2]
```diff
@@ -28,6 +25,8 @@ pub struct LinearMemory {
 
 /// It holds the raw bytes of memory accessed by a WebAssembly Instance
 impl LinearMemory {
+    pub const PAGE_SIZE: u32 = 65536;
+    pub const MAX_PAGES: u32 = 65536;
     pub const WASM_PAGE_SIZE: usize = 1 << 16; // 64 KiB
```
[the remark this answers is not in the document]
Comment (appcypher) [reply src/webassembly/memory.rs]: True. I'm only leaving them there because their values may be different in the future.

**Comment by appcypher** [comment]
Makes sense. I will keep that in mind once I figure out how to represent/store the runtime data.

**Comment by syrusakbary** [comment]
While we acknowledge having Nginx is still a work in progress, we feel we are super close to have it running. So in order to don't diverge too much from master, we will merge this PR and open a separate one for the final steps for running nginx.

**Pull request merged**